res/controllers/Pioneer-DDJ-400-script.js

```diff
@@ -0,0 +1,741 @@
+    }
+
+    // trigger "track loaded" animations when a track is loaded
+    engine.makeConnection("[Channel1]", "track_loaded", PioneerDDJ400.trackLoadedLED);
+    engine.makeConnection("[Channel2]", "track_loaded", PioneerDDJ400.trackLoadedLED);
+
+    // eye candy : play the "track loaded" animation on both decks at startup
+    midi.sendShortMsg(0x9F, 0x00, 0x7F);
+    midi.sendShortMsg(0x9F, 0x01, 0x7F);
+
+    // resets pad mode to hotcue
+    PioneerDDJ400.performancePads.reset();
+
+    // turn on loop in and out lights
+    PioneerDDJ400.setLoopButtonLights(0x90, 0x7F);
+    PioneerDDJ400.setLoopButtonLights(0x91, 0x7F);
+
+    // handle loop toggle events
+    engine.makeConnection("[Channel1]", "loop_enabled", PioneerDDJ400.loopToggle);
+    engine.makeConnection("[Channel2]", "loop_enabled", PioneerDDJ400.loopToggle);
+
+    // poll the controller for current control positions on startup
+    midi.sendSysexMsg([0xF0, 0x00, 0x40, 0x05, 0x00, 0x00, 0x02, 0x06, 0x00, 0x03, 0x01, 0xf7], 12);
+};
+
+//
+// Channel level lights
+//
+
+PioneerDDJ400.vuMeterUpdate = function(value, group) {
+    var newVal = value * 150;
+
+    switch (group) {
+    case "[Channel1]":
+        midi.sendShortMsg(0xB0, 0x02, newVal);
+        break;
+
+    case "[Channel2]":
+        midi.sendShortMsg(0xB1, 0x02, newVal);
+        break;
+    }
+};
+
+//
+// Tempo sliders
+//
+
+PioneerDDJ400.tempoSliderMSB = function(channel, control, value, status, group) {
+    PioneerDDJ400.highResMSB[group].tempoSlider = value;
+};
+
+PioneerDDJ400.tempoSliderLSB = function(channel, control, value, status, group) {
+    var fullValue = (PioneerDDJ400.highResMSB[group].tempoSlider << 7) + value;
+
+    engine.setValue(
+        group,
+        "rate",
+        ((0x4000 - fullValue) - 0x2000) / 0x2000
+    );
+};
+
+//
+// Effects
+//
+
+PioneerDDJ400.numFxSlots = 3;
+
+Object.defineProperty(PioneerDDJ400, "selectedFxSlot", {
+    get: function() {
+        return engine.getValue("[EffectRack1_EffectUnit1]", "focused_effect");
+    },
+    set: function(value) {
+        if (value < 0 || value > PioneerDDJ400.numFxSlots) {
+            return;
+        }
+        engine.setValue("[EffectRack1_EffectUnit1]", "focused_effect", value);
+        var isEffectEnabled = engine.getValue(PioneerDDJ400.selectedFxGroup, "enabled");
+        PioneerDDJ400.toggleLight(PioneerDDJ400.lights.beatFx, isEffectEnabled);
+    },
+});
+
+Object.defineProperty(PioneerDDJ400, "selectedFxGroup", {
+    get: function() {
+        return "[EffectRack1_EffectUnit1_Effect" + PioneerDDJ400.selectedFxSlot + "]";
+    },
+});
+
+PioneerDDJ400.beatFxLevelDepthRotate = function(_channel, _control, value) {
+    var newVal = value === 0 ? 0 : (value / 0x7F);
+
+    if (engine.getValue(PioneerDDJ400.selectedFxGroup, "enabled")) {
+        engine.softTakeoverIgnoreNextValue(PioneerDDJ400.selectedFxGroup, "meta");
+        engine.setValue("[EffectRack1_EffectUnit1]", "mix", newVal);
```

Comment:
```suggestion
        engine.setParameter("[EffectRack1_EffectUnit1]", "mix", value / 0x7F);
```
This allows for removing the `newVal` variable.

Reply:
nice spot :slightly_smiling_face: